File: CSharp/SimpleWatchdog/SimpleWatchdog.cs
// *** SimpleWatchdog ***

// This example shows the usage of the watchdog for communication over virtual serial port and Bluetooth,
// 
// - Use bluetooth connection
// - Use auto scanning and connecting
// - Use watchdog 

using System;
using CommandMessenger;
using CommandMessenger.Serialport;
using CommandMessenger.TransportLayer;

namespace SimpleWatchdog
{
    class SimpleWatchdog
    {
        enum Command
        {
            Identify,          // Command to identify device
            Acknowledge        // Command to acknowledge a received command
        };

        public bool RunLoop { get; set; }

        private const string UniqueDeviceId = "BFAF4176-766E-436A-ADF2-96133C02B03C";
        private static ITransport _transport;
        private static CmdMessenger _cmdMessenger;
        private static ConnectionManager _connectionManager;

        // Setup function
        public void Setup()
        {
            _transport = new SerialTransport {CurrentSerialSettings = {DtrEnable = false}};
                // some boards (e.g. Sparkfun Pro Micro) DtrEnable may need to be true.                        
            // We do not need to set serial port and baud rate: it will be found by the connection manager                                                           

            // Initialize the command messenger with the Serial Port transport layer
            _cmdMessenger = new CmdMessenger(_transport)
            {
                BoardType = BoardType.Bit16, // Set if it is communicating with a 16- or 32-bit Arduino board
                PrintLfCr = false // Do not print newLine at end of command, to reduce data being sent
            };

            _cmdMessenger.Attach(OnUnknownCommand);
            _cmdMessenger.Attach((int) Command.Acknowledge, OnAcknowledge);

            // We don't need to provide a handler for identify command - this is a job for Connection Manager.
            _connectionManager = new SerialConnectionManager((_transport as SerialTransport), _cmdMessenger,
                (int) Command.Identify, UniqueDeviceId)
            {
                // Enable watchdog functionality.
                WatchdogEnabled = true,

                // By default connection settings are persisted. 
                // In this way, when the application is restarted the previously succesfull settings are first tried
                //PersistentSettings = false,

                // Instead of scanning for the connected port, you can also use a fixed port. Set this port through the CurrentSerialSettings
                //UseFixedPort = false
            };

            // Show all connection progress on command line             
            _connectionManager.Progress += (sender, eventArgs) =>
            {
                // If you want to reduce verbosity, you can only show events of level 1 or 2
                if (eventArgs.Level <= 3) Console.WriteLine(eventArgs.Description);
            };
        
            // Finally - activate connection manager
            _connectionManager.StartConnectionManager();
        }

        // Loop function
        public void Loop()
        {
            //Wait for key
            Console.ReadKey();            
            // Stop loop
            RunLoop = false;  
        }

        // Exit function
        public void Exit()
        {
            _connectionManager.Dispose();
            _cmdMessenger.Disconnect();
            _cmdMessenger.Dispose();
            _transport.Dispose();
        }

        static void OnUnknownCommand(ReceivedCommand command)
        {
            // Add your handling of unknown commands here
        }

        static void OnAcknowledge(ReceivedCommand command)
        {
            // Add your handling of command acknowlegdement here
        }
    }
}

File: extras/Python/samples/simple_watchdog/simple_watchdog.py
"""SimpleWatchdog sample — auto-connect and supervise a device.

Pairs with ``examples/SimpleWatchdog/SimpleWatchdog.ino``. The sketch responds
to the identify command (id 0) with its unique id string.
"""
from __future__ import annotations

import time

from cmd_messenger import (
    BoardType,
    CmdMessenger,
    ConnectionManagerProgressEventArgs,
    SerialConnectionManager,
)
from cmd_messenger.transport.serial import SerialSettings, SerialTransport

BAUD = 115200
UNIQUE_DEVICE_ID = "BFAF4176-766E-436A-ADF2-96133C02B03C"  # match the sketch
IDENTIFY_CMD_ID = 0


def on_progress(args: ConnectionManagerProgressEventArgs) -> None:
    print(f"  [{args.level}] {args.description}")


def main() -> None:
    transport = SerialTransport(SerialSettings(baud_rate=BAUD))
    messenger = CmdMessenger(transport, board_type=BoardType.BIT_16)
    messenger.print_lf_cr = True

    connmgr = SerialConnectionManager(
        transport,
        messenger,
        watchdog_command_id=IDENTIFY_CMD_ID,
        unique_device_id=UNIQUE_DEVICE_ID,
    )
    connmgr.progress += on_progress
    connmgr.connection_found += lambda: print("** Device found **")
    connmgr.connection_timeout += lambda: print("** Device timed out **")
    connmgr.watchdog_enabled = True

    try:
        connmgr.start_connection_manager()
        # Run for 30 seconds.
        time.sleep(30)
    except KeyboardInterrupt:
        pass
    finally:
        connmgr.stop_connection_manager()
        messenger.dispose()


if __name__ == "__main__":
    main()
